### PR TITLE
add SharpEdges article

### PR DESCRIPTION
**Pull request by JaapWijnen**

### CODE REVIEW

**Comment by JaapWijnen** [comment]
@BradLarson Added @porterchild as co author

File: Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md [hunk 1]
```diff
@@ -7,11 +7,142 @@ This is an overview of some of the currently missing capabilities in differentia
...
+- `for..in` loops currently require you to loop over indices and using `withoutDerivative(at:)` on the collection you're looping over.
+- `map()` and `reduce()` cannot be differentiated through directly, instead there are special functions for `differentiableMap()` and `differentiableReduce()`.
 - Differentiation through `_modify` subscript accessors (like array subscript setters) isn't supported yet.
 - Differentiation through `@_alwaysEmitIntoClient` tagged functions isn’t yet supported. The most common cases of these are in SIMD functions, like `.sum()`.
-- `map()` and `reduce()` cannot be differentiated through directly, instead there are special functions for `differentiableMap()` and `differentiableReduce()`.
 - No support yet in the standard library for Dictionary differentiation.
-- `for..in` loops currently require using `withoutDerivative(at:)` on the collection you're looping over.
+- FloatingPoint type conversions are not differentiable out of the box. Workarounds are available however.
+- No support yet for differentiable keypath subscripting (get or set) but workarounds are available.
```
Comment (BradLarson) [inline Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]: Some of these bullet-point topics are expanded upon below, and others are only mentioned within this bullet list. Would it make sense to remove the topics that are covered below, and move this to the bottom after the detailed issues as an "Other issues" section?

File: Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md [hunk 2]
```diff
@@ -7,11 +7,142 @@ This is an overview of some of the currently missing capabilities in differentia
...
+
+
+## Examples
```
Comment (BradLarson) [inline Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]: My preference would be to remove the "Examples" title and let each of the following issue sections stand on their own.

File: Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md [hunk 3]
```diff
@@ -7,11 +7,142 @@ This is an overview of some of the currently missing capabilities in differentia
...
+### Map and Reduce
+The `map` and `reduce` methods do not currently support closures marked with `@differentiable` but there's special differentiable versions of these that work exactly like you're used to:
+```swift
+let a = [1.0, 2.0, 3.0]
+let aPlusOne = a.differentiableMap { $0 + 1.0 } // [2.0, 3.0, 4.0]
+let aSum = a.differentiableReduce { 0, + } // 6.0
+```
+
+### Array subscript setters (`_modify` subscript accessors)
+Currently the subcript setters on arrays (`array[0] = 1.0`) are not differentiable. Under the hood this is due to `_modify` subscript accessors not supporting differentiability yet (This feature is hopefully landing in Swift 5.10)
```
Comment (BradLarson) [inline Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]:
```suggestion
Currently the subcript setters on arrays (`array[0] = 1.0`) are not differentiable. Under the hood this is due to `_modify` subscript accessors not supporting differentiability yet. (Work is ongoing, and this feature should land in Swift soon.)
```

File: Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md (hunk 1 above)
Comment (JaapWijnen) [reply Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]: Yeah I pictured having a quick overview of all items and then expanding on those in the sections below. But I'm also in favour of this approach.

Comment (JaapWijnen) [reply Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]: Found a better way! Links to the sections at the top and an Other section at the bottom. Let me know what you think

Comment (BradLarson) [reply Sources/DifferentiableSwiftExamplesDocumentation/DifferentiableSwiftExamples.docc/SharpEdgesInDifferentiableSwift.md]: Yeah, I like how that turned out. Thanks for the reformat, looks great!

**Review by BradLarson — APPROVED** [review]

**Pull request merged**